astropy/units/distribution.py

```diff
@@ -0,0 +1,427 @@
+
+
+class Distribution:
+    """
+    A unitful value with associated uncertainty distribution.  While subclasses
+    may have exact analytic forms, this class uses samples to represent the
+    distribution.
+
+    Parameters
+    ----------
+    distr : array-like
+        The distribution, with sampling along the *leading* axis.
+    distr_center : `None` or array-like
+        The "center" of this distribution.  It must have the same shape as
+        ``distr``, aside from the (initial) sample dimension. If `None`, the
+        *median* of ``distr``  along the sampling axis will be used.
+    unit : astropy unit
+        A unit of the same sort that `~astropy.units.Quantity` accepts.
+
+    Additional argumnets are passed into the `~astropy.units.Quantity`
+    constructor.
+
+    Notes
+    -----
+    This class has an attribute ``stat_view`` which can be set to an
+    `numpy.ndarray` subclass to determine what the summary statistics are
+    converted to.  By default this is `~astropy.units.Quantity`, but it is
+    available so that subclasses can change it.
+    """
+
+    def __new__(cls, distr, distr_center=None, unit=None):
+        if isinstance(distr, Distribution):
+            if distr_center is None:
+                distr_center = distr.distr_center
+            distr = distr.distribution
+        else:
+            distr = np.asanyarray(distr, order='C')
+        if distr.shape == ():
+            raise TypeError('Attempted to initialize a Distribution with a scalar')
+
+        # Not clear whether one wants the distr_center argument still.
+        # Would need __array_finalize__ to preserve it through slicing, etc.
+        if distr_center is None:
+            distr_center = np.median(distr, axis=-1)
+
+        # Not clear why the unit argument is needed.
+        if unit is not None:
+            distr = Quantity(distr, unit, copy=False, subok=True)
+
+        new_dtype = np.dtype({'names': ['samples'],
+                              'formats': [(distr.dtype, (distr.shape[-1],))]})
+        distr_cls = type(distr)
+        if not issubclass(distr_cls, Distribution):
+            # Probably should have a dict on the class with these, so
+            # we don't create new classes needlessly.
+            new_name = distr_cls.__name__ + cls.__name__
+            new_cls = type(new_name, (cls, distr_cls),
+                           {'_distr_cls': distr_cls})
+        self = distr.view(dtype=new_dtype, type=new_cls)
+        # Get rid of trailing dimension of 1.
+        self.shape = distr.shape[:-1]
+        self.center = distr_center
+        return self
+
+    @property
+    def distribution(self):
+        return self['samples']
+
+    def __getitem__(self, item):
+        result = super().__getitem__(item)
+        if item == 'samples':
+            return super(Distribution, result).view(self._distr_cls)
+        else:
+            return Distribution.__new__(self.__class__, result['samples'])
```

Comment:
Why is this necessary?  I would think you can just `return result` here - it should already be a `Distribution` with the correct unit (if this is to redefine `center`, then all the more reason to get rid of it!!)